invertedai/logs/common.py

```diff
@@ -0,0 +1,456 @@
+    """
+    A class for containing features relevant to both log reading and writing such as visualization.
+    """
+
+    def __init__(self):
+        self.scenario_log = None
+        self.simulation_length = None
+
+    def visualize_up_to_timestep(
+        self,
+        timestep,
+        gif_path,
+        fov = 200,
+        resolution = (2048,2048),
+        dpi = 300,
+        map_center = None,
+        direction_vec = False,
+        velocity_vec = False,
+        plot_frame_number = True
+    ):
+        assert timestep >= 1 or timestep <= (self.simulation_length - 1), "Simulation period to plot is invalid."
+
+        location_info_response = location_info(location=self.scenario_log.location)
+        rendered_static_map = location_info_response.birdview_image.decode()
+        map_center = tuple([location_info_response.map_center.x, location_info_response.map_center.y]) if map_center is None else map_center
+
+        scene_plotter = ScenePlotter(
+            map_image=rendered_static_map,
+            fov=fov,
+            xy_offset=map_center,
+            static_actors=location_info_response.static_actors,
+            resolution=resolution,
+            dpi=dpi
+        )
+        scene_plotter.initialize_recording(
+            agent_states=self.scenario_log.agent_states[0],
+            agent_properties=self.scenario_log.agent_properties,
+            traffic_light_states=None if self.scenario_log.traffic_lights_states is None else self.scenario_log.traffic_lights_states[0]
+        )
+
+        traffic_lights_states = [None]*(timestep-1) if self.scenario_log.traffic_lights_states is None else self.scenario_log.traffic_lights_states[1:timestep]
+        for states, lights in zip(self.scenario_log.agent_states[1:timestep],traffic_lights_states):
+            scene_plotter.record_step(states,lights)
+
+        fig, ax = plt.subplots(constrained_layout=True, figsize=(50, 50))
+        plt.axis('off')
+        scene_plotter.animate_scene(
+            output_name=gif_path,
+            ax=ax,
+            direction_vec=direction_vec,
+            velocity_vec=velocity_vec,
+            plot_frame_number=plot_frame_number,
+        )
+
+    def visualize(
+        self,
+        gif_path,
+        fov = 200,
+        resolution = (2048,2048),
+        dpi = 300,
+        map_center = None,
+        direction_vec = False,
+        velocity_vec = False,
+        plot_frame_number = True
+    ):
+        self.visualize_up_to_timestep(
+            timestep = self.simulation_length-1,
+            gif_path = gif_path,
+            fov = fov,
+            resolution = resolution,
+            dpi = dpi,
+            map_center = map_center,
+            direction_vec = direction_vec,
+            velocity_vec = velocity_vec,
+            plot_frame_number = plot_frame_number
+        )
+
+class LogWriter(LogHandler):
+    """
+    A class for conveniently writing a log to a JSON log format.
+    """
+
+    def __init__(self):
```

Comment:
You don't need to declare it if no explicit modification is needed

Reply:
It theoretically doesn't make a difference in this instance but it does initialize the 2 instance parameters. My main reasoning however is that if changes are made in the future or by a user, then they don't need to worry about adding this line to make sure the LogWriter class is initialized properly.